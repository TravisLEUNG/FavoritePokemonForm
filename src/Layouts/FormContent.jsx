import { useState, useEffect } from "react";
import { _hyphen2Label, _camel2Label } from "../Tools/StringCaseTransformation";
import _ from "lodash";
import Alert from "react-bootstrap/Alert";

import FormTextField from "../Components/FormTextField";
import FormSelectionField from "../Components/FormSelectionField";
import ButtonField from "../Components/ButtonField";
import TableView from "../Components/TableView";

import { contentData } from "../Data/FormContentData";

const validation = (format, value) => (!_.isEmpty(value) && format.test(value));
const objectValueUpdate = (prevObj, index, value) => ({
  ...prevObj,
  [index]: value,
});

export default function FormContent(props) {
  const [pokemonList, setPokemonList] = useState([]);
  const [fieldInputs, setFieldInputs] = useState(
    () =>
      props.data || {
        FirstName: "",
        LastName: "",
        PhoneNumber: "+",
        EmailAddress: "",
        FavoritePokemon: {},
      }
  );
  const [pokemonDetails, setPokemonDetails] = useState({
    name: "",
    abilities: [],
    height: "0",
    weight: "0",
    types: [],
  });
  const [hasAlert, setHasAlert] = useState(false);
  const [fieldErrorMsg, setFieldErrorMsg] = useState({});
  useEffect(() => {
    async function getPokemon() {
      const res = await fetch(
        "https://pokeapi.co/api/v2/pokemon/?offset=0&limit=1154"
      );
      const data = await res.json();
      setPokemonList(
        data.results.map((item) => ({
          label: _hyphen2Label(item.name),
          value: item.name,
          url: item.url,
        }))
      );
    }
    getPokemon();
  }, []);

  useEffect(() => {
    props.onDataChange(fieldInputs);

    async function getPokemonDetail() {
      const res = await fetch(fieldInputs["FavoritePokemon"].url);
      const { name, abilities, height, weight, types } = await res.json();
      setPokemonDetails({ name, abilities, height, weight, types });
    }
    getPokemonDetail();
  }, [props, fieldInputs]);

  /* Supporting Func */
  const validInput = (field, value) => {
    setFieldInputs((prevFieldInputs) =>
      objectValueUpdate(prevFieldInputs, field, value)
    );
    setFieldErrorMsg((prevFieldErrorMsg) =>
      objectValueUpdate(prevFieldErrorMsg, field, "")
    );
  };
  const invalidInput = (field, errorMsg) => {
    setFieldErrorMsg((prevFieldErrorMsg) =>
      objectValueUpdate(prevFieldErrorMsg, field, errorMsg)
    );
  };

  /* Event Func */
  const onChange = ({ field, value }) => {
    if (validation(contentData[field].format, value)) {
      validInput(field, value);
    } else {
      invalidInput(field, contentData[field].errorMsg);
    }
  };

  const onSelect = ({ field, option }) => {
    if (validation(contentData[field].format, option.value)) {
      validInput(field, option);
    } else {
      invalidInput(field, contentData[field].errorMsg);
    }
  };

  const onSubmit = (event) => {
    event.preventDefault();
    for (const [field, value] of Object.entries(fieldInputs)) {
      const checkingValue = _.isEqual(field, "FavoritePokemon")
        ? value.value
        : value;
      if (
        _.isEmpty(checkingValue) ||
        !validation(contentData[field].format, checkingValue)
      ) {
        setHasAlert(true);
        setTimeout(() => {
          setHasAlert(false);
        }, 2000);
        return false;
      }
    }
    props.handleViewChange("Review");
  }

  /* Data formatting */
  const propsList = Object.keys(contentData).reduce((obj, field) => {
    obj[field + "Props"] = {
      ...contentData[field],
      label: _camel2Label(contentData[field].fieldId),
      handleChange: onChange,
      msg: fieldErrorMsg[field],
      value: fieldInputs[field],
    };
    return obj;
  }, {});
  const pokemonDetailRows = [
    { header: "Pokemon", data: _hyphen2Label(pokemonDetails.name) },
    {
      header: "Abilities",
      data: pokemonDetails.abilities
        .map((a) => _hyphen2Label(a.ability.name))
        .join(", "),
    },
    { header: "Height", data: `${pokemonDetails.height} m` },
    { header: "Weight", data: `${pokemonDetails.weight} kg` },
    {
      header: "Types",
      data: pokemonDetails.types
        .map((t) => _hyphen2Label(t.type.name))
        .join(", "),
    },
  ];

  return (
    <form
      className="form-content"
    >
      <Alert
        className="form-content_alert"
        key="FormIncomplete"
        variant="danger"
        show={hasAlert}
      >
        Form is not completed yet!
      </Alert>

      <h3>Let us know your Favorite Pokemon !</h3>
      <p>Fill in the form and share you pokemon preferences.</p>

      <div className="form-content_input">
        {/* Name input field */}
        <FormTextField {...propsList["FirstNameProps"]} />
        <FormTextField {...propsList["LastNameProps"]} />
        {/* Phone number input field */}
        <FormTextField {...propsList["PhoneNumberProps"]} />
        {/* Email address input field */}
        <FormTextField {...propsList["EmailAddressProps"]} />
        {/* Favorite Pokemon selection field */}
        <FormSelectionField
          {...propsList["FavoritePokemonProps"]}
          optionList={pokemonList}
          handleChange={onSelect}
        />
        <TableView rowList={pokemonDetailRows} />
      </div>

      <div className="d-grid gap-3 form-content_button">
        <ButtonField
          label="Next"
          isSubmit
          handleClick={onSubmit}
        />
      </div>
    </form>
  );
}
