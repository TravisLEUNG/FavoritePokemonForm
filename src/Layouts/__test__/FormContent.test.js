import { render, screen, cleanup, waitFor } from "@testing-library/react";
import user from "@testing-library/user-event"
import renderer from "react-test-renderer"

import FormContent from "../FormContent";

const fakeData = {
  FirstName: "",
  LastName: "",
  PhoneNumber: "",
  EmailAddress: "",
  FavoritePokemon: {},
};
const fakeCompletedData = {
  FirstName: "Ultimate",
  LastName: "Hokon",
  PhoneNumber: "+19876543210",
  EmailAddress: "ultimate.hokon@gmail.com",
  FavoritePokemon: {
    label: "Venusaur",
    url: "https://pokeapi.co/api/v2/pokemon/3/",
    value: "venusaur",
  },
};
const fakeEvent = () => {};
describe('FormContent', () => {
  const onDataChange = jest.fn();
  const onViewChange = jest.fn();

  afterEach(() => {
    cleanup();
  });

  it('Go to Review page when all fields pass validation', async () => {
    render(<FormContent data={{...fakeData, FavoritePokemon: fakeCompletedData["FavoritePokemon"]}} onDataChange={onDataChange} handleViewChange={onViewChange}/>);

    // 1st step
    user.type(getPhoneNumber(), "+1");
    expect(onDataChange).toHaveBeenCalledTimes(3);
    user.type(getFirstName(), fakeCompletedData["FirstName"]);
    expect(onDataChange).toHaveBeenCalledTimes(3+fakeCompletedData["FirstName"].length);
    user.type(getLastName(), "T");
    expect(onDataChange).toHaveBeenCalledTimes(4+fakeCompletedData["FirstName"].length);
    user.type(getEmailAddress(), "T");
    expect(onDataChange).toHaveBeenCalledTimes(5+fakeCompletedData["FirstName"].length);

    // 2nd step
    clickReviewButton();
    expect(onViewChange).toHaveBeenCalledTimes(1);
  });

  it('Has alert when required fields empty', async () => {
    render(<FormContent data={{...fakeData, FavoritePokemon: fakeCompletedData["FavoritePokemon"]}} onDataChange={fakeEvent} handleViewChange={onViewChange}/>);

    clickReviewButton();
    expect(onViewChange).toHaveBeenCalledTimes(0);
    await waitFor(() => {
      expect(getAlertContent()).toBeInTheDocument();
    });
  });

  it('Has error message when fields input wrongly', async () => {
    render(<FormContent data={{...fakeData, FavoritePokemon: fakeCompletedData["FavoritePokemon"]}} onDataChange={fakeEvent} handleViewChange={fakeEvent}/>);

     user.type(getFirstName(), "!");
     user.type(getLastName(), "!");
     user.type(getPhoneNumber(), "!");
     user.type(getEmailAddress(), "!");
    
    expect(screen.getAllByText(/sorry, only letters \(a-z,a-z\) are allowed\./i)).toHaveLength(2);
    expect(screen.getByText(/sorry, only numbers \(0-9\) are allowed and start with sign \(\+\) with country code./i)).toBeInTheDocument();
    expect(screen.getByText(/sorry, only letters \(a\-z\), numbers \(0\-9\), and signs \(\.\-_@\) are allowed\./i)).toBeInTheDocument();
  });
})

test("Match snapshot", () => {
  const tree = renderer
    .create(<FormContent data={{...fakeData, FavoritePokemon: fakeCompletedData["FavoritePokemon"]}} onDataChange={fakeEvent} handleViewChange={fakeEvent}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

function getFirstName () {
  return screen.getByRole('textbox', {
    name: /First Name/i
  });
}
function getLastName () {
  return screen.getByRole('textbox', {
    name: /Last Name/i
  });
}
function getPhoneNumber () {
  return screen.getByRole('textbox', {
    name: /Phone Number/i
  });
}
function getEmailAddress () {
  return screen.getByRole('textbox', {
    name: /Email Address/i
  });
}

function getAlertContent() {
  return screen.getByText(/Form is not completed yet!/i)
}

function clickReviewButton() {
  return user.click(screen.getByRole('button', { name: /Next/i }));
}