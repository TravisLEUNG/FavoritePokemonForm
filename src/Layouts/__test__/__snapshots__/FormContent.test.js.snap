// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
<form
  className="form-content"
>
  <h3>
    Let us know your Favorite Pokemon !
  </h3>
  <p>
    Fill in the form and share you pokemon preferences.
  </p>
  <div
    className="form-content_input"
  >
    <div
      className="form-text-field"
    >
      <label
        className="form-label"
        htmlFor="FirstName"
      >
        First name
      </label>
      <input
        className="form-control"
        id="FirstName"
        name="FirstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your first name"
        value=""
      />
      <small
        className="error-msg form-text"
        id="FirstNameHelpBlock"
      />
    </div>
    <div
      className="form-text-field"
    >
      <label
        className="form-label"
        htmlFor="LastName"
      >
        Last name
      </label>
      <input
        className="form-control"
        id="LastName"
        name="LastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your last name"
        value=""
      />
      <small
        className="error-msg form-text"
        id="LastNameHelpBlock"
      />
    </div>
    <div
      className="form-text-field"
    >
      <label
        className="form-label"
        htmlFor="PhoneNumber"
      >
        Phone number
      </label>
      <input
        className="form-control"
        id="PhoneNumber"
        name="PhoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your phone number"
        value=""
      />
      <small
        className="error-msg form-text"
        id="PhoneNumberHelpBlock"
      />
    </div>
    <div
      className="form-text-field"
    >
      <label
        className="form-label"
        htmlFor="EmailAddress"
      >
        Email address
      </label>
      <input
        className="form-control"
        id="EmailAddress"
        name="EmailAddress"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your email address"
        value=""
      />
      <small
        className="error-msg form-text"
        id="EmailAddressHelpBlock"
      />
    </div>
    <div
      className="form-selection-field"
    >
      <label
        className="form-label"
        htmlFor="FavoritePokemon"
      >
        Favorite pokemon
      </label>
      <div
        className="basic-single css-b62m3t-container"
        id="FavoritePokemon"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-5-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
        />
        <div
          className="select__control css-1t8y0t9-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="select__value-container select__value-container--has-value css-319lph-ValueContainer"
          >
            <div
              className="select__single-value css-qc6sy-singleValue"
            >
              Venusaur
            </div>
            <div
              className="select__input-container css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-expanded={false}
                aria-haspopup={true}
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className="select__input"
                disabled={false}
                id="react-select-5-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="select__indicators css-1hb7zxy-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              className="select__indicator select__loading-indicator css-at12u2-loadingIndicator"
            >
              <span
                className="css-1xtdfmb-LoadingDot"
              />
              <span
                className="css-zoievk-LoadingDot"
              />
              <span
                className="css-x748d8-LoadingDot"
              />
            </div>
            <span
              className="select__indicator-separator css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="FavoritePokemon"
          type="hidden"
          value="venusaur"
        />
      </div>
      <small
        className="error-msg form-text"
        id="FavoritePokemonHelpBlock"
      />
    </div>
    <div
      className="table-view"
    >
      <table>
        <tbody>
          <tr>
            <th>
              Pokemon
              : 
            </th>
            <td />
          </tr>
          <tr>
            <th>
              Abilities
              : 
            </th>
            <td />
          </tr>
          <tr>
            <th>
              Height
              : 
            </th>
            <td>
              0 m
            </td>
          </tr>
          <tr>
            <th>
              Weight
              : 
            </th>
            <td>
              0 kg
            </td>
          </tr>
          <tr>
            <th>
              Types
              : 
            </th>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="d-grid gap-3 form-content_button"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Next
    </button>
  </div>
</form>
`;
